pkg/model/components/cloudcontrollermanager.go

```diff
@@ -0,0 +1,123 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package components
+
+import (
+	"fmt"
+
+	//"github.com/golang/glog"
+	"k8s.io/kops/pkg/apis/kops"
+	"k8s.io/kops/pkg/apis/kops/util"
+	"k8s.io/kops/upup/pkg/fi"
+	"k8s.io/kops/upup/pkg/fi/loader"
+)
+
+type CloudControllerManagerOptionsBuilder struct {
+	Context *OptionsContext
+}
+
+var _ loader.OptionsBuilder = &CloudControllerManagerOptionsBuilder{}
+
+func (b *CloudControllerManagerOptionsBuilder) BuildOptions(o interface{}) error {
+
+	clusterSpec := o.(*kops.ClusterSpec)
+	if clusterSpec.CloudControllerManager == nil {
+		clusterSpec.CloudControllerManager = &kops.CloudControllerManagerConfig{}
+	}
+	ccm := clusterSpec.CloudControllerManager
+
+	if !clusterSpec.EnableExternalCloudController {
+		return nil
+	}
+
+	k8sv17, err := util.ParseKubernetesVersion("v1.7.0")
+	if err != nil {
+		return fmt.Errorf("Unable to parse kubernetes version %s", err)
+	}
+
+	kubernetesVersion, err := KubernetesVersion(clusterSpec)
+	if err != nil {
+		return fmt.Errorf("Unable to parse kubernetesVersion %s", err)
+	}
+
+	if !kubernetesVersion.GTE(*k8sv17) {
+		return fmt.Errorf("Kubernetes version should be >=1.7.0 to enable cloud controller manager")
+	}
+
+	switch kops.CloudProviderID(clusterSpec.CloudProvider) {
+	case kops.CloudProviderAWS:
+		ccm.CloudProvider = "aws"
+	case kops.CloudProviderGCE:
+		ccm.CloudProvider = "gce"
+	case kops.CloudProviderVSphere:
+		ccm.CloudProvider = "vsphere"
```

Comment:
do?

Reply:
This is actually a case where it would break for DO and [any other cloud providers not in kubernetes core](https://github.com/kubernetes/kubernetes/tree/master/pkg/cloudprovider/providers) right now. Here we're assuming that `EnableExternalCloudController` will always create a `CloudControllerManagerConfig` in cluster spec which actually only works for the cloud providers that are bulit into `cloud-controller-manager`.  I think a better approach in general is to have cloud controller managers as a cluster addon as opposed to adding more component configs in the kops API.

Reply:
fwiw, DO support is still alpha and this is more of a future proofing concern.